src/main/java/org/benetech/servicenet/adapter/healthleads/persistence/PersistanceManager.java

```diff
@@ -0,0 +1,198 @@
+            HealthleadsLanguage.class, externalServiceId, HealthleadsService.class));
+    }
+
+    private Set<Location> getLocationsToPersist(String organizationId) {
+        Set<HealthleadsLocation> locations = dictionary.getRelatedEntities(
+            HealthleadsLocation.class, organizationId, HealthleadsOrganization.class);
+        Set<Location> result = new HashSet<>();
+        for (HealthleadsLocation healthleadsLocation : locations) {
+            String externalLocationId = healthleadsLocation.getId();
+
+            Location location = mapper.extractLocation(healthleadsLocation);
+
+            location.setProviderName(PROVIDER_NAME);
+            location.setPhysicalAddress(getPhysicalAddressesToPersist(externalLocationId));
+            location.setPhones(getPhonesForLocationToPersist(externalLocationId));
+            location.setLangs(getLanguagesForLocationToPersist(externalLocationId));
+            result.add(location);
+        }
+        return result;
+    }
+
+    private Set<Language> getLanguagesForLocationToPersist(String externalLocationId) {
+        return mapper.extractLanguages(dictionary.getRelatedEntities(
+            HealthleadsLanguage.class, externalLocationId, HealthleadsLocation.class));
+    }
+
+    private Set<Phone> getPhonesForServiceToPersist(String externalServiceId) {
+        return mapper.extractPhones(dictionary.getRelatedEntities(
+            HealthleadsPhone.class, externalServiceId, HealthleadsService.class));
+    }
+
+    private Set<Phone> getPhonesForLocationToPersist(String externalLocationId) {
+        return mapper.extractPhones(dictionary.getRelatedEntities(
+            HealthleadsPhone.class, externalLocationId, HealthleadsLocation.class));
+    }
+
+    //TODO: What if there are multiple addresses
+    private PhysicalAddress getPhysicalAddressesToPersist(String locationId) {
+        Set<HealthleadsPhysicalAddress> addresses = dictionary.getRelatedEntities(
+            HealthleadsPhysicalAddress.class, locationId, HealthleadsLocation.class);
+        if (addresses.isEmpty()) {
+            return null;
+        }
+        return mapper.extractPhysicalAddress(addresses.iterator().next());
+    }
+
+    private Set<ServiceTaxonomy> getTaxonomiesToPersist(String serviceId) {
+        Set<HealthleadsServiceTaxonomy> serviceTaxonomies = dictionary.getRelatedEntities(
+            HealthleadsServiceTaxonomy.class, serviceId, HealthleadsService.class);
+
+        Set<ServiceTaxonomy> result = new HashSet<>();
+        for (HealthleadsServiceTaxonomy healthleadsServiceTaxonomy : serviceTaxonomies) {
+            ServiceTaxonomy serviceTaxonomy = mapper.extractServiceTaxonomy(healthleadsServiceTaxonomy)
+                .providerName(PROVIDER_NAME).externalDbId(serviceId);
+            Set<HealthleadsTaxonomy> taxonomies = dictionary.getRelatedEntities(
+                HealthleadsTaxonomy.class, healthleadsServiceTaxonomy.getTaxonomyId(), HealthleadsBaseData.class);
```

Comment:
Why did you use HealthleadsBaseData class instead of HealthleadsServiceTaxonomy class?

Reply:
It should be done this way as all data that are not Location/Organization/Service related, are just considered to be HealthleadsBaseData related (line 54) as specifying it precisely would bring no benefits here.